DigitalLearningSolutions.Data.Tests/Services/ActionPlanServiceTests.cs

```diff
@@ -0,0 +1,110 @@
+        {
+            // Given
+            const int competencyLearningResourceId = 1;
+            const int delegateId = 2;
+            const int selfAssessmentId = 3;
+            const string resourceName = "Activity";
+            const string resourceLink = "www.test.com";
+            const int learningLogId = 4;
+            const int learningHubResourceId = 6;
+
+            var addedDate = new DateTime(2021, 11, 1);
+            A.CallTo(() => clockService.UtcNow).Returns(addedDate);
+
+            var competencyLearningResource = new CompetencyLearningResource
+            {
+                LearningHubResourceReference = learningHubResourceId,
+            };
+            A.CallTo(
+                () => competencyLearningResourcesDataService.GetCompetencyLearningResourceById(
+                    competencyLearningResourceId
+                )
+            ).Returns(competencyLearningResource);
+
+            A.CallTo(() => learningHubApiService.GetResourceNameAndLink(learningHubResourceId))
+                .Returns((resourceName, resourceLink));
+
+            var resourceCompetencies = new[] { 1, 2, 3, 4, 5, 6, 7, 8 };
+            A.CallTo(
+                () => competencyLearningResourcesDataService.GetCompetencyIdsByLearningHubResourceReference(
+                    learningHubResourceId
+                )
+            ).Returns(resourceCompetencies);
+
+            var assessmentCompetencies = new[] { 2, 3, 5, 6, 8, 9, 10 };
+            A.CallTo(() => selfAssessmentDataService.GetCompetencyIdsForSelfAssessment(selfAssessmentId))
+                .Returns(assessmentCompetencies);
+
+            A.CallTo(
+                () => learningLogItemsDataService.InsertLearningLogItem(
+                    A<int>._,
+                    A<DateTime>._,
+                    A<string>._,
+                    A<string>._,
+                    A<int>._
+                )
+            ).Returns(learningLogId);
+
+            // When
+            actionPlanService.AddResourceToActionPlan(competencyLearningResourceId, delegateId, selfAssessmentId);
+
+            // Then
+            A.CallTo(
+                () => learningLogItemsDataService.InsertLearningLogItem(
+                    delegateId,
+                    addedDate,
+                    resourceName,
+                    resourceLink,
+                    competencyLearningResourceId
+                )
+            ).MustHaveHappenedOnceExactly();
+            A.CallTo(
+                () => learningLogItemsDataService.InsertCandidateAssessmentLearningLogItem(
+                    selfAssessmentId,
+                    learningLogId
+                )
+            ).MustHaveHappenedOnceExactly();
+            A.CallTo(
+                () => learningLogItemsDataService.InsertLearningLogItemCompetencies(learningLogId, A<int>._, addedDate)
+            ).MustHaveHappened(5, Times.Exactly);

```

Comment:
At the risk of making this test even larger, what do you think of additionally adding a similar set of assertions for this to test the competency Ids are the correct ones, such as (excuse the formatting):

```
 var expectedMatchingCompetencies = new[] { 2, 3, 5, 6, 8 };
            foreach (var competencyId in expectedMatchingCompetencies)
            {
                A.CallTo(
                    () => learningLogItemsDataService.InsertLearningLogItemCompetencies(learningLogId, competencyId, addedDate)
                ).MustHaveHappenedOnceExactly(); 
            }
```
Would probably need expectedMatchingCompetencies to be defined alongside the other collections of competencies somewhere obvious.